bucketlist/app.py

```diff
@@ -0,0 +1,281 @@
+        '''Ensures that get method is not allowed for registration'''
+        return({'message': 'Method is not allowed'}, 405)
+
+
+class LoginApi(Resource):
+    '''Logs in a user into the api. Give  the user a token to perfom transactions as authentication.'''
+
+    def post(self):
+        parser = reqparse.RequestParser()
+        parser.add_argument('username', required=True,
+                            help="Username cannot be blank")
+        parser.add_argument('password', required=True,
+                            help="Password cannot be blank")
+        credentials = parser.parse_args()
+        uname = credentials['username']
+        pword = credentials['password']
+
+        # confirm that the user does exist
+        user = User.query.filter_by(username=uname).first()
+        if user is None:
+            return({'message': 'user does not exist'}, 404)
+        else:
+            # verify the password
+            if user.veriry_password(pword, user.password):
+                token = user.generate_token()
+                return({'message': 'login successful',
+                        'Authorization': token.decode('ascii')}, 200)
+            else:
+                return({'message': 'wrong username/password combination'}, 401)
+
+    def get(self):
+        '''Ensures that get method not allowed in logging in'''
+        return({'message': 'Method is not allowed'}, 405)
+
+
+class BucketLists(Resource):
+    '''Adds and retrieves bucketlists to and from the database'''
+    @auth.login_required
+    def get(self):
+        # Adds page and limit variables to be used for pagination
+        page = request.args.get('page', 1)
+        limit = request.args.get('limit', 1) or 100
+        # Adds a query attribute specified in the url
+        q = request.args.get('q')
+        if q:
+            # Searches bucketlists for one with the given query
+            buckets = Bucketlist.query.filter(Bucketlist.bucketlist_name.ilike(
+                '%' + q + "%")).filter_by(created_by=g.user.id).paginate(int(page), int(limit), False)
+            bkts = buckets.items
+            if buckets:
+                if buckets.has_next:
+                    next_page = str(
+                        request.url_root) + 'bucketlist/api/bucketlists?q=' + q + '&page=' + str(int(page) + 1)
+                else:
+                    next_page = 'None'
+
+                if buckets.has_prev:
+                    prev = str(request.url_root) + 'bucketlist/api/bucketlists?q=' + \
+                        q + '&page=' + str(int(page) - 1)
+                else:
+                    prev = 'None'
+                buckets = [bucket for bucket in bkts]
+                response = {'bucketlists': marshal(
+                    buckets, Bucketlist_marshaller),
+                    'next': next_page,
+                    'prev': prev}
+                return response
+            else:
+                return {'message': ' Bucketlist not found'}, 404
```

Comment:
I decided to create the api in such a way that any search is paginated - whether a query search or getting all the bucketlists.